### ISSUE TITLE
QNX 7.0 LLVM cxx bugfix

### ISSUE THREAD

**Issue by Johnnyxy** [boostorg/config#288]
Hi there,
I get serval compiler errors when compiling Boost source 1.70.0 (currently available version) as shared/static libraries with QNX SDP 7.0 (GCC 5.4.0).

This has been discussed here too: https://github.com/boostorg/asio/issues/248#issuecomment-498182528

output
--------
See attached output file:
[boost_qnx_70_cxx_error.txt](https://github.com/boostorg/config/files/3302507/boost_qnx_70_cxx_error.txt)

the error
----------
The error is a combination of code and compiler configuration (through compiler header files).
Boost uses a feature that gets disabled when using the LLVM C++ library what results in the compiler errors.
Actually when using a QCC target that uses the GNU C++ library ```gpp``` everything works fine only with the LLVM C++ library ```cxx``` this error occurs.

error origin
------------
The problem is located in the header file ```target/qnx7/usr/include/c++/v1/__config``` in the ```#ifdef block``` at line 409 (or with current QNX SDP 7.0 update at line 412):
>#if defined(__GNUC__)

In this block the macro ```_LIBCPP_HAS_NO_TEMPLATE_ALIASES``` gets defined at line 432 (or with current QNX SDP 7.0 update at line 435):
>#define _LIBCPP_HAS_NO_TEMPLATE_ALIASES

This is actually wrong as the GCC compiler supports the [template alias](https://gcc.gnu.org/projects/cxx-status.html#cxx11) feature since version 4.7.

workaround 1: works with every user code
-------------------------------------------
One can circumvent the problem by undefining the macro after the ```__config``` file gets included.
But then one has to ensure the header inclusion order in the code what could be tricky sometimes.

It is possible to circumvent the problem without changing the compiler headers, code or introducing additional macro arguments.

1. create a file named ```__config```
2. put that file in a path which is accessible from the code project (readable)
3. write the following code into that file:
```c
// This is a workaround for a problem with QNX 7 / qcc 5.4 and LLVM C++ library.
// E.g. boost cannot be compiled without this as with the LLVM C++ library qcc
// disables several compiler features which is a misconfiguration of the
// compiler settings.

#ifndef __QCC-5-4_CXX___CONFIG_66771214-7934-4DDC-94A3-B0511E80AE69__
#define __QCC-5-4_CXX___CONFIG_66771214-7934-4DDC-94A3-B0511E80AE69__

#include_next <__config>
#if defined(__QNXNTO__) && \
    (defined(__cpp_alias_templates) && (__cpp_alias_templates >= 200704L)) || \
    (defined(__GNUC__) && ((__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 7))))
#    undef _LIBCPP_HAS_NO_TEMPLATE_ALIASES
#endif

#endif // __QCC-5-4_CXX___CONFIG_66771214-7934-4DDC-94A3-B0511E80AE69__
```
4. put the path to this file into the header search paths
    e.g. ```q++ ... -isystem c:/path/to_the_file```
5. compile the code normally without any changes

What happens here is that a feature of GCC/QCC is used [include_next](https://gcc.gnu.org/onlinedocs/cpp/Wrapper-Headers.html) ([see docs here](https://gcc.gnu.org/onlinedocs/cpp/Wrapper-Headers.html)) to "override" the inclusion of the header file globally (in the context of the compilation) and instead include the custom one.

To make things easier see the following attachment:
[__config.zip](https://github.com/boostorg/config/files/3305678/__config.zip)

workaround 2: works with every user code
-------------------------------------------
This workaround is not recommended as it will only work with the specifically modified header files of the compiler installation.
Manipulate the compiler header file ```target/qnx7/usr/include/c++/v1/__config``` and comment the problematic line ```#define _LIBCPP_HAS_NO_TEMPLATE_ALIASES``` to ```// #define _LIBCPP_HAS_NO_TEMPLATE_ALIASES```.

**Comment by jzmaddock** [comment]
This isn't a Boost.Config issue: we already define BOOST_NO_CXX11_ALLOCATOR in exactly this situation, the issue is that ASIO isn't using Boost.Config so you will have to take the matter up with them.

Likewise the patches referenced above are scattered over multiple libraries: please submit a separate issue for each library patched so the appropriate maintainers can review them.

**Comment by evadeflow** [comment]
@Johnnyxy: *Thank you* so much, this information was an absolute life-saver for me, almost five years after you posted it. I just used your *workaround 1* to get an app to compile for QNX7. I had to fight some battles to get permission to use Asio, and  convinced my colleagues by showing them a demo app I compiled on Ubuntu. The Boost headers were *already* in our cross-build, including the Asio headers, so I assumed things would *Just Work*&trade;. I was sweating quite a bit when the compiler *exploded* at me the first time I tried to cross-compile. Until I found your workaround, I was worried that perhaps I shouldn't have opened my big mouth about how Asio would make this task *so* much easier.

Sorry for spamming the issue tracker, just wanted to let you know that your efforts *years ago* are still paying dividends to strangers like me&mdash;thank you!
